Make various improvements to OptionButton

aekobear opened the pull request:
<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

Fixed several bugs from https://github.com/godotengine/godot/issues/6558

- OptionButton can now be set to no selection by calling `optionButton.select(-1)`
- The `Selected` property in the editor can now be set to `-1` for no selection
  - The `Selected` property can now be restored to default (which is -1)
- When a selected item is deleted, the selection is set to no selection (-1)
  - This also fixes the scenarios where all items are deleted and where items are added after being deleted
  
 I also found an undocumented bug introduced here: https://github.com/godotengine/godot/pull/54647
 
 The PopupMenu had been edited to use IDs starting at 0 in some places, but in other places (and in OptionButton) it was expected to use IDs starting at 1. This caused a bug where when adding OptionButton items from the inspector, both of the first two items would have ID = 1
 
 Rather than reverting, I fully converted PopupMenu and OptionButton to use 0-based IDs. This matches the documentation, which claims that IDs are set by index (which starts with 0). It also improves some dissonance where "0" could be passed in as a valid object ID but was also being used to represent "no selection" when calling `get_selected_id()`. Now, -1 represent no selection in both the index and the ID flows.

akien-mga commented on scene/gui/option_button.cpp:
```diff
@@ -234,6 +236,10 @@ Ref<Texture2D> OptionButton::get_item_icon(int p_idx) const {
 }
 
 int OptionButton::get_item_id(int p_idx) const {
+  if (p_idx == NONE_SELECTED) {
+    return NONE_SELECTED;
+  }
```
Note that you should use tabs for indentation.

akien-mga replied:
And after/while fixing that, please squash commits together by doing a rebase and force push (see [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase)).

eikobear replied:
fixed and squashed!

akien-mga reviewed (approved):
Looks good to me. I think it would be worth mentioning the new behavior in the `select` docs too:
```
Passing `[code]-1[/code]` as index deselects any currently selected item.
```

aekobear committed:
Make various improvements to OptionButton

- Allow OptionButton selection to be set to -1 to signify no selection, both via API and in the editor.
- Reset OptionButton selection to -1 when the selected item has been removed.
- Fully convert PopupMenu to a zero-based ID system, which improves an inconsistency in generated IDs when making new items in the editor.

eikobear commented:
> Looks good to me. I think it would be worth mentioning the new behavior in the `select` docs too:
> 
> ```
> Passing `[code]-1[/code]` as index deselects any currently selected item.
> ```

Good point- updated

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:
